src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+
+/// Represents the result of parsing the args of a nailgunnable ExecuteProcessRequest
+/// TODO We may want to split the classpath by the ":", and store it as a Vec<String>
+///         to allow for deep fingerprinting.
+struct ParsedArgLists {
+    nailgun_args: Vec<String>,
+    client_args: Vec<String>,
+}
+
+///
+/// Given a list of args that one would likely pass to a java call,
+/// we automatically split it to generate two argument lists:
+///  - nailgun arguments: The list of arguments needed to start the nailgun server.
+///    These arguments include everything in the arg list up to (but not including) the main class.
+///    These arguments represent roughly JVM options (-Xmx...), and the classpath (-cp ...).
+///
+///  - client arguments: The list of arguments that will be used to run the jvm program under nailgun.
+///    These arguments can be thought of as "passthrough args" that are sent to the jvm via the nailgun client.
+///    These arguments include everything starting from the main class.
+///
+/// We assume that:
+///  - Every args list has a main class.
+///  - There is exactly one argument that doesn't begin with a `-` in the command line before the main class,
+///    and it's the value of the classpath (i.e. `-cp scala-library.jar`).
+///
+/// We think these assumptions are valid as per: https://github.com/pantsbuild/pants/issues/8387
+///
+fn split_args(args: &Vec<String>) -> ParsedArgLists {
+    let mut nailgun_args = vec![];
+    let mut client_args = vec![];
+    let mut have_seen_classpath = false;
+    let mut have_processed_classpath = false;
+    let mut have_seen_main_class = false;
+    for arg in args {
+        if have_seen_main_class {
+            // Arguments to pass to the client request
+            client_args.push(arg.clone());
+        } else if (arg == "-cp" || arg == "-classpath") && !have_seen_classpath {
+            // Process the -cp flag itself
+            have_seen_classpath = true;
+            nailgun_args.push(arg.clone());
+        } else if have_seen_classpath && !have_processed_classpath {
+            // Process the classpath string
+            nailgun_args.push(arg.clone());
+            have_processed_classpath = true;
+        } else if have_processed_classpath && !arg.starts_with("-") {
+            // Process the main class:
+            // I have already seen the value of the -cp classpath, and this is not a flag.
+            client_args.push(arg.clone());
+            have_seen_main_class = true;
+        } else {
+            // Process the rest (jvm options to pass to nailgun)
+            nailgun_args.push(arg.clone());
+        }
+    }
+    ParsedArgLists {
+        nailgun_args: nailgun_args,
+        client_args: client_args
+    }
+}
+
+// TODO(8481) The input_files arg should be calculated from deeply fingerprinting the classpath.
+fn get_nailgun_request(args: Vec<String>, _input_files: Digest, jdk: Option<PathBuf>) -> ExecuteProcessRequest {
```

Comment:
Can you add some docs about what this function does? (There are two "nailgun" requests - one which starts the server, one which runs the client, it's not clear which of them this one is from the name)